Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            g_variant_builder_add(&options, "{sv}", "token", g_variant_new_uint32(*preparePrintResponse->token()));
+
+            // FIXME: title
+            GRefPtr<GVariant> arguments(g_variant_new("(ssha{sv})", "", _("Print"), fdIndex, &options));
+
+            g_dbus_proxy_call_with_unix_fd_list(priv->portalProxy.get(), "Print", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, fdList.get(), NULL,
+                [](GObject* proxy, GAsyncResult* result, gpointer userData) {
+                    auto printOperation = adoptGRef(static_cast<WebKitPrintOperation*>(userData));
+                    GUniqueOutPtr<GError> error;
+                    GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(proxy), result, &error.outPtr()));
+                    if (error)
+                        webkitPrintOperationFailed(printOperation.get(), WTFMove(error));
+                    webkitPrintOperationFinished(printOperation.get());
+                }, printOperation.leakRef());
+        }, printOperation.leakRef(), nullptr);
+    });
+}
+
+static WebKitPrintOperationResponse webkitPrintOperationRunPortalDialog(WebKitPrintOperation* printOperation, GtkWindow* parent)
+{
+    auto* priv = printOperation->priv;
+
+    if (!priv->portalProxy) {
+        GUniqueOutPtr<GError> error;
+
+        priv->portalProxy = adoptGRef(g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_NONE, nullptr,
+            "org.freedesktop.portal.Desktop", "/org/freedesktop/portal/desktop", "org.freedesktop.portal.Print",
+            nullptr, &error.outPtr()));
+
+        // Fallback to the unix dialog on error
+        if (error)
+            return WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL;
+    }
+
+    RELEASE_ASSERT(priv->portalProxy);
+
+    auto* connection = g_dbus_proxy_get_connection(priv->portalProxy.get());
+    auto uniqueName = String::fromUTF8(g_dbus_connection_get_unique_name(connection));
+    auto senderName = makeStringByReplacingAll(uniqueName.substring(1), '.', '_');
+    auto token = makeString("WebKitGTK", weakRandomNumber<uint32_t>());
+    auto requestPath = makeString("/org/freedesktop/portal/desktop/request/", senderName, "/", token);
+
+    RELEASE_ASSERT(!priv->signalId);
+
+    priv->signalId = g_dbus_connection_signal_subscribe(connection, "org.freedesktop.portal.Desktop", "org.freedesktop.portal.Request",
+        "Response", requestPath.ascii().data(), nullptr, G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+        [](GDBusConnection* connection, const char*, const char*, const char*, const char*, GVariant* parameters, gpointer userData) {
+            WebKitPrintOperation* printOperation = static_cast<WebKitPrintOperation*>(userData);
+            auto* priv = printOperation->priv;
+            GRefPtr<GVariant> options;
+            uint32_t response;
+
+            RELEASE_ASSERT(priv->signalId);
+            g_dbus_connection_signal_unsubscribe(connection, priv->signalId);
```

Comment:
Hm, I'm worried this line could destroy the print operation and `priv` along with it, because it will call the GDestroyNotify, right? I think you need a protector ref here. Instead of:

```
WebKitPrintOperation* printOperation = static_cast<WebKitPrintOperation*>(userData);
```

You would do:

```
GRefPtr<WebKitPrintOperation> protectedPrintOperation(static_cast<WebKitPrintOperation*>(userData));
```

Note the lack of `adoptRef()`. Alternatively you could omit the GDestroyNotify and adopt the ref instead, but that would probably make it harder to read rather than easier.

P.S. This code is relatively tricky, with so many callbacks, plus lack of cancellation support throughout making things even harder, that I'm actually pleased I didn't find _more_ memory safety errors. So good job overall.